src/browser/modules/Sidebar/ProjectFiles.tsx

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright (c) 2002-2020 "Neo4j,"
+ * Neo4j Sweden AB [http://neo4j.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+import React, { useEffect, useState } from 'react'
+import { connect } from 'react-redux'
+import { withBus } from 'react-suber'
+import { Bus } from 'suber'
+import { useMutation } from '@apollo/client'
+
+import Render from 'browser-components/Render'
+import { Drawer, DrawerHeader } from 'browser-components/drawer'
+import ProjectFilesScripts, { ProjectFilesError } from './ProjectsFilesScripts'
+import {
+  StyledSetting,
+  StyledSettingLabel,
+  StyledSettingTextInput
+} from './styled'
+import {
+  setProjectFileDefaultFileName,
+  updateCacheAddProjectFile
+} from './project-files.utils'
+import { SET_CONTENT } from 'shared/modules/editor/editorDuck'
+import { CYPHER_FILE_EXTENSION } from 'shared/services/export-favorites'
+import {
+  SELECT_PROJECT_FILE,
+  SAVE_PROJECT_FILE,
+  ADD_PROJECT_FILE,
+  PROJECT_FILES_MOUNTED,
+  PROJECT_FILES_UNMOUNTED
+} from './project-files.constants'
+
+interface ProjectFiles {
+  bus: Bus
+  projectId: string
+}
+
+const ProjectFiles = ({ bus, projectId }: ProjectFiles) => {
+  const [addFile, { error: apolloError }] = useMutation(ADD_PROJECT_FILE)
+  const [isSaveMode, setSaveMode] = useState(false)
+  const [fileName, setFileName] = useState('')
+  const [fileContents, setFileContents] = useState('')
+  const [error, setError] = useState('')
+
+  useEffect(() => {
+    let isStillMounted = true
+    // when ProjectFileButton is clicked in save mode
+    bus &&
+      bus.take(SAVE_PROJECT_FILE, editorValue => {
+        if (isStillMounted) {
+          setSaveMode(true)
+          setFileName(setProjectFileDefaultFileName(editorValue))
+          setFileContents(editorValue)
+          setError('')
+        }
+      })
+    // when a non-Project File action sets content in the editor
+    bus &&
+      bus.take(SET_CONTENT, () => {
+        if (isStillMounted) {
+          setSaveMode(false)
+          setFileName('')
+          setFileContents('')
+          setError('')
+        }
+      })
+    return () => {
+      isStillMounted = false
+    }
+  }, [])
+
+  useEffect(() => {
+    // send mounted/unmounted message
+    bus && bus.send(PROJECT_FILES_MOUNTED, '')
+    return () => bus && bus.send(PROJECT_FILES_UNMOUNTED, '')
+  }, [])
```

Comment:
don't you need to pass `bus` here?

Reply:
probably. Ill double check all of them

Reply:
looking into this, its recommended by React to not use empty dependency arrays anyway especially when there are dependencies in the effect - they recommend [exhaustive-deps](https://github.com/facebook/react/issues/14920 ) lint rule (as part of  eslint-plugin-react-hooks). So I will change my usage of the dependency array everywhere it makes sense now and create a card for the linting (I tried and it raised an unrelated issue I dont have time to look into right now)

Reply:
Ah yes that's what I meant. Really the only thing that matters is that without the second parameter it gets called every render